src/Avalonia.Base/DirectPropertyMetadata`1.cs

```diff
@@ -46,6 +46,6 @@ public override void Merge(AvaloniaPropertyMetadata baseMetadata, AvaloniaProper
             }
         }
 
-        public override AvaloniaPropertyMetadata GenerateTypeSafeMetadata() => new DirectPropertyMetadata<TValue>(UnsetValue, DefaultBindingMode, EnableDataValidation);
+        public override AvaloniaPropertyMetadata GenerateTypeSafeMetadata() => this;
```

Comment:
For `DirectPropertyMetadata` the change doesn't really matter, it was made for consistency with `StyledPropertyMetadata`.

However the equivalent change in `StyledPropertyMetadata` is one of the core optimization here for `AvaloniaProperty.GetMetadata()`. With the changes in this PR, for most properties `_singleMetadata == _defaultMetadata` (except when there's a `CoerceCallback`) and we can get through the fast path without extra lookups. Reverting that would completely remove the optimization.

Note that while making the changes for this PR, I checked that all code paths in the Avalonia codebase for `AvaloniaPropertyMetadata.Merge()` were only for `OverrideMetadata`/`AddOwner`, effectively making the metadata immutable after the initial merge.

Somehow, I completely missed that `Merge` was public and callable on any existing metadata!

It means default metadata can be modified _after_ a property is defined, which is really an unsupported scenario and should probably be blocked with an exception. For example, this PR also caches the default value if there's a single common one, and won't ever notice that the default value for a metadata has been changed through `Merge`.

With metadata properly frozen the `Merge`-after-initialization scenario doesn't exist anymore, and we can keep optimizations in place were needed.

(It will be technically a breaking change to throw in `Merge()`; in practice anyone modifying the default metadata for already defined properties is introducing bugs - let's hope that nobody is doing that currently).

Reply:
@TomEdwardsEnscape I've opened #15384 to make the property metadata immutable after initialization, so the changes in this PR would become valid.